src/mesh_tools.rs

```diff
@@ -1,12 +1,200 @@
+use std::collections::{HashMap, HashSet, VecDeque};
 
 use nalgebra::base::Vector3;
 use nalgebra::geometry::Point3;
 use smallvec::{smallvec, SmallVec};
 
 use crate::convert::{cast_u32, cast_usize};
-use crate::geometry::{Face, Geometry, TriangleFace};
-use crate::mesh_topology_analysis::face_to_face_topology;
+use crate::geometry::{Face, Geometry, OrientedEdge, TriangleFace, UnorientedEdge};
+use crate::mesh_topology_analysis;
+
+/// Orients all the faces the same way - matches their winding (vertex order).
+///
+/// This function crawls the mesh geometry and flips all the faces, which are
+/// not facing the same way as the previously checked neighboring faces,
+/// starting with the first face in the list for each mesh island.
+///
+/// The algorithm relies on the fact that in a proper manifold mesh, each
+/// oriented edge has exactly one (for watertight mesh geometry) or none (for
+/// mesh patch naked edges) counter-edge oriented the opposite direction. It
+/// crawls the geometry and if a face neighboring the current one doesn't have
+/// the proper winding, it's being reverted and only then triggers checking its
+/// own neighbors.
+///
+/// This method doesn't flip the normals associated with the face vertices, as
+/// there is no unambiguous way to do so automatically.
+///
+/// https://gist.github.com/vTurbine/16fbb99225ad4c0ac80b24855dd61a7c
+///
+/// # Warning
+/// As a result, the entire mesh can end up facing inwards (be entirely
+/// reverted). At the moment we have no tools to detect such a case
+/// automatically, so we need to rely on the user to check it and potentially
+/// revert winding of the entire mesh.
+///
+/// # Warning
+/// The results might be unpredictable for non-manifold meshes and moebius-like
+/// topologies.
+///
+// FIXME: Flip also vertex normals if the visual/practical tests prove it's
+// needed
+#[allow(dead_code)]
+pub fn synchronize_mesh_winding(
+    geometry: &Geometry,
+    oriented_edges: &[OrientedEdge],
+    unoriented_edges: &[UnorientedEdge],
+    face_to_oriented_edge_topo: &[SmallVec<[u32; 3]>],
+    unoriented_edge_to_face_topo: &[SmallVec<[u32; 2]>],
+) -> Geometry {
+    // Processing queue: indices of faces sharing edges with the current face,
+    // zipped with the oriented edge they should contain if they have a proper
+    // winding.
+    let mut face_indices_with_desired_oriented_edge_queue: VecDeque<(u32, OrientedEdge)> =
+        VecDeque::new();
+    let mut visited = vec![false; geometry.faces_len()];
+    let mut proper_faces: Vec<TriangleFace> = Vec::with_capacity(geometry.faces_len());
+
+    // lookup table
+    let oriented_to_unoriented_edge_index: Vec<_> = oriented_edges
+        .iter()
+        .map(|o_e| {
+            let to_unoriented = o_e.to_unoriented();
+            unoriented_edges
+                .iter()
+                .position(|u_e| *u_e == to_unoriented)
+                .expect("No unoriented edge matches the oriented")
+        })
+        .collect();
```

Comment:
True and not deduplicating `UnorientedEdge`s is a mistake. Should be fixed. It is quadratic but the calculation would be needed anyway - either collected into a data structure like this or calculated as a logic where needed (and every item of the list will be needed).
As I mentioned above, it would be way more efficient if calculated together with `OrientedEdge`s and `UnorientedEdge`s. Then we could define a new topology type containing all three of these structures.